src/lib/template-factory.ts

```diff
@@ -44,19 +44,40 @@ export type TemplateFactory = (result: TemplateResult) => Template;
 export function templateFactory(result: TemplateResult) {
   let templateCache = templateCaches.get(result.type);
   if (templateCache === undefined) {
-    templateCache = new Map<TemplateStringsArray, Template>();
+    templateCache = new Map<string, Template>();
     templateCaches.set(result.type, templateCache);
   }
-  let template = templateCache.get(result.strings);
+  let key = templateKeys.get(result.strings);
+  if (key !== undefined) {
+    // If templateKeys has the key it's guaranteed to exist in the templateCache
+    return templateCache.get(key)!;
+  }
+
+  // If the result.strings are new, generate a key and set it in templateKeys
+  key = result.strings.join(marker);
+  templateKeys.set(result.strings, key);
+
+  let template = templateCache.get(key);
   if (template === undefined) {
     template = new Template(result, result.getTemplateElement());
-    templateCache.set(result.strings, template);
+  } else {
+    // If a template already exists, refresh the key object in the templateCache
+    templateCache.delete(key);
```

Comment:
Don't think you need this `delete` because you are setting it again which would overwrite the previous value.

Reply:
I'm actually unsure here. We're setting to a value which is equal to, but not identical to, the previous value. I bet that Map implementations are free to replace the key or not in that case, as aside from the perf we're trying to eek out, there shouldn't be an observable difference.

Reply:
I already tested this. The delete is necessary to force the key to update. The benchmark I linked above tests a version that sets after deleting, and without deleting, and the version that does not delete does not update the key object.